### ISSUE TITLE
Optimizer parameters 

### ISSUE THREAD

**Issue by Stanwang1210** [espnet/espnet#5133]
**Describe the bug**
Hi! I'm wondering whether this is a bug or not.
Take a look at the [https://github.com/espnet/espnet/blob/fc37f80ff96070107c61a2a020f9627f82d646c5/espnet2/tasks/abs_task.py#L911](https://github.com/espnet/espnet/blob/fc37f80ff96070107c61a2a020f9627f82d646c5/espnet2/tasks/abs_task.py#L911) 

Due to [Issue](https://discuss.pytorch.org/t/parameters-with-requires-grad-false-are-updated-during-training/90096/2)
Should this line be revised into optim = optim_class([p for p in model.parameters() if p.requires_grad], **args.optim_conf)

If this line needs to be revised, then maybe another revision is needed here [https://github.com/espnet/espnet/blob/fc37f80ff96070107c61a2a020f9627f82d646c5/espnet2/optimizers/optim_groups.py#L56](https://github.com/espnet/espnet/blob/fc37f80ff96070107c61a2a020f9627f82d646c5/espnet2/optimizers/optim_groups.py#L56)

**Comment by Fhrozen** [comment]
Don't worry about this, 
If you check for the date, that issue was reported in 2020 and now is no longer an issue.
I also tested because of some issues I had no long ago, but that does not matter.
You can try it by adding the lines you mentioned and storing the weights, then compare them and see if there is any change.

**Comment by rishabh135** [comment]
Hey @Fhrozen  thanks for this update, can I ask a relevant related query ? 

I was trying to run the CTC model with different learning rate and as indicated in  [espnet/espnet2/tasks/abs_task.py](https://github.com/espnet/espnet/blob/fc37f80ff96070107c61a2a020f9627f82d646c5/espnet2/tasks/abs_task.py#L883) 

with my modified param_groups in optimizer:

```
    @classmethod
    def build_optimizers(
        cls,
        args: argparse.Namespace,
        model: torch.nn.Module,
    ) -> List[torch.optim.Optimizer]:

        logging.warning("  asr_lr {} att_lr {} ctc_lr {}  \n".format(args.asr_lr, args.att_lr, args.ctc_lr))
        if cls.num_optimizers != 1:
            raise RuntimeError(
                "build_optimizers() must be overridden if num_optimizers != 1"
            )

        optim_class = optim_classes.get(args.optim)
        if optim_class is None:
            raise ValueError(f"must be one of {list(optim_classes)}: {args.optim}")

        if(args.ngpu > 1):
            param_grp = [
                {'params': model.module.encoder.parameters(), 'lr': args.asr_lr},
                {'params': model.module.decoder.parameters(), 'lr': args.att_lr},
                {'params': model.module.ctc.parameters(), 'lr': args.ctc_lr},
            ]
        else:
            param_grp = [
                {'params': model.encoder.parameters(), 'lr': args.asr_lr},
                {'params': model.decoder.parameters(), 'lr': args.att_lr},
                {'params': model.ctc.parameters(), 'lr': args.ctc_lr},
                ]

        optimi = torch.optim.Adam((param_grp), betas=(0.9,0.999),weight_decay=0.00001)

        optimizers = [optimi]
        return optimizers
```
The training falters and suddenly collapses even with the default value of 0.002 for asr_lr, ctc_lr and att_lr, with NaNs, can you please help me identify what's that I am missing?

**Comment by Fhrozen** [comment]
Did you try with:
```python
param_grp = [
                {'params': model.encoder.parameters()},
                {'params': model.decoder.parameters(), 'lr': args.att_lr},
                {'params': model.ctc.parameters(), 'lr': args.ctc_lr},
                ]

optimi = torch.optim.Adam(param_grp, lr=args.asr_lr, betas=(0.9,0.999), weight_decay=0.00001)
```

Also, check the weight_decay; it seems a little higher. Try with 1e-8, or 0

**Comment by Fhrozen** [comment]
@rishabh135 
TDLR: I found no issue using the param group in the optimizer.

Installed a fresh espnet clone using:
- torch=1.13.1
- CUDA=11.7
- recipe: egs2/jsut/asr1 for fast debugging.
- model: Conformer8
- GPU: RTX3060 x 2

added the following lines in `espnet2/tasks/asr.py` (I recommed to use the task related instead of the abs one, so other tasks does not get affected)

```
@classmethod
    def build_optimizers(
        cls,
        args: argparse.Namespace,
        model: torch.nn.Module,
    ) -> List[torch.optim.Optimizer]:
        if cls.num_optimizers != 1:
            raise RuntimeError(
                "build_optimizers() must be overridden if num_optimizers != 1"
            )

        optim_class = optim_classes.get(args.optim)
        if optim_class is None:
            raise ValueError(f"must be one of {list(optim_classes)}: {args.optim}")
        if args.sharded_ddp:
            raise NotImplementedError
        else:
            
            if args.exclude_weight_decay:
                optim = configure_optimizer(
                    model,
                    optim_class,
                    args.optim_conf,
                    args.exclude_weight_decay_conf,
                )
            else:
                optim_conf = args.optim_conf
                if "lr_group" in optim_conf:
                    lr_grp = optim_conf["lr_group"]
                    del optim_conf["lr_group"]
                else:
                    _lr = optim_conf["lr"]
                    lr_grp = [_lr, _lr, _lr]

                param_grp = [
                    {"params": model.encoder.parameters(), "lr": lr_grp[0]},
                    {"params": model.decoder.parameters(), "lr": lr_grp[1]},
                    {"params": model.ctc.parameters(), "lr": lr_grp[2]},
                ]

                optim = optim_class(param_grp, **optim_conf)

        optimizers = [optim]
        return optimizers
```

I tested using the same value in the lr_grp, and using different values for each part.
similar values:

the trainer displays similar initial_lr for each part:
```
Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    capturable: False
    differentiable: False
    eps: 1e-08
    foreach: None
    fused: False
    initial_lr: 4.0
    lr: 6.324555320336758e-08
    maximize: False
    weight_decay: 0

Parameter Group 1
    amsgrad: False
    betas: (0.9, 0.999)
    capturable: False
    differentiable: False
    eps: 1e-08
    foreach: None
    fused: False
    initial_lr: 4.0
    lr: 6.324555320336758e-08
    maximize: False
    weight_decay: 0

Parameter Group 2
    amsgrad: False
    betas: (0.9, 0.999)
    capturable: False
    differentiable: False
    eps: 1e-08
    foreach: None
    fused: False
    initial_lr: 4.0
    lr: 6.324555320336758e-08
    maximize: False
    weight_decay: 0
)
```

and the figs:

![sm_cer](https://github.com/espnet/espnet/assets/11988996/5b702a74-f37e-40ed-bbd8-96c818678d2f)
![sm_loss](https://github.com/espnet/espnet/assets/11988996/1ed8e8bb-471f-473b-be5f-54ff35b25a3b)
![sm_optim0_lr0](https://github.com/espnet/espnet/assets/11988996/8427d70e-1aea-4301-b2e2-21d2645cf602)
![sm_optim0_lr1](https://github.com/espnet/espnet/assets/11988996/1a34b952-7705-496f-af0b-8ed60f9162a4)
![sm_optim0_lr2](https://github.com/espnet/espnet/assets/11988996/8833022f-ef5b-4d2d-bd05-71a9bf323b36)

shows correct training

for different values:

![cer](https://github.com/espnet/espnet/assets/11988996/16c38265-ea5c-41e3-bb2f-565650c2cdff)
![loss](https://github.com/espnet/espnet/assets/11988996/a5832535-2bb4-48e1-aacb-a75ed6afb7e3)
![optim0_lr0](https://github.com/espnet/espnet/assets/11988996/0be6e6e3-43bf-4a39-99a3-b696fce5e1ca)
![optim0_lr1](https://github.com/espnet/espnet/assets/11988996/992ef02a-f602-4de0-b471-c506ffa86e15)
![optim0_lr2](https://github.com/espnet/espnet/assets/11988996/5eb891f5-7a00-4eab-a94d-abdc69cc63f2)

No NaNs displayed, and the training seems smooth.
So not sure whether is the installation, the model, or the recipe you are using.

Try to include additional details, so I can test.
